### DISCUSSION TITLE
Skia.Gtk project throw error with CommunityToolkit.Mvvm 8

### DISCUSSION THREAD

**Discussion by mcNets** [unoplatform/uno#9486] [General]
After adding the CommunityToolkit.Mvvm 8 to my project it throwed an error building the Skia.Gtk project. Windows project builds and runs without any problem.

Error message:

>UXAML0001	Unable to find member [SelectedItem] on type [test_community.MainPageViewModel]	test_community.Skia.Gtk	F:\Projectes\Test\uno-winui\test_community\test_community\test_community.Skia.Gtk\CSC	1	

This error appears after I added this line to the xaml code:

`SelectedItem="{x:Bind ViewModel.SelectedItem, Mode=TwoWay}"`

It seems that is not able to use the proerty SelectedItem auto-generated by the community toolkit.

I let the instructions to reproduce it.

Create a new uno-platform project for WinUI & Skia.Gtk and name it test_community.

Add the CommunityTookit.Mvvm (8.0) to both projects.

**Add a ItemModel class**
```
namespace test_community
{
    public class ItemModel
    {
        public int Id { get; set; }
        public string Name { get; set; }
        public override string ToString() => $"{Id}-{Name}";
    }
}
```

**MainPage.xaml**
```
<Page
    x:Class="test_community.MainPage"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:test_community"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    mc:Ignorable="d"    
    Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">

    <StackPanel>
        <ListView ItemsSource="{x:Bind ViewModel.Items}"
                  SelectedItem="{x:Bind ViewModel.SelectedItem, Mode=TwoWay}"/>
    </StackPanel>
</Page>
```

**MainPage.xaml.cs**
```
using Microsoft.UI.Xaml.Controls;

namespace test_community
{
    public sealed partial class MainPage : Page
    {
        public MainPageViewModel ViewModel = new MainPageViewModel();

        public MainPage()
        {
            this.InitializeComponent();
            this.DataContext = ViewModel;
        }
    }
}
```

**Add a MainPageViewModel class**
```
using CommunityToolkit.Mvvm.ComponentModel;
using System.Collections.ObjectModel;

namespace test_community
{
    [ObservableObject]
    public partial class MainPageViewModel
    {
        public ObservableCollection<ItemModel> Items { get; set; }

        [ObservableProperty]
        private ItemModel _selectedItem;

        public MainPageViewModel()
        {
            Items = new ObservableCollection<ItemModel>()
            {
                new ItemModel() {Id = 1, Name = "Item 1"},
                new ItemModel() {Id = 2, Name = "Item 2"},
                new ItemModel() {Id = 3, Name = "Item 3"},
            };
        }
    }
}
```

**Comment by yoshiask** [comment]
It seems like the source generator isn't being run on the Skia head. Is `MainPageViewModel` in the shared project?

**Reply by mcNets** [reply]
Yes, it is

**Reply by yoshiask** [reply]
Have you made sure to add a reference to `CommunityToolkit.Mvvm` to the Skia head?

**Reply by mcNets** [reply]
Yes, but as soon as I get my laptop I will paste the project conf of both projects.

**Comment by mcNets** [comment]
ItemModel and MainPageViewModel are in the shared project.

**Comment by HavenDV** [comment]
Try to change
```cs
[ObservableProperty]
private ItemModel _selectedItem;
```
to
```cs
[ObservableProperty]
private ItemModel selectedItem;
```
And
```cs
public MainPageViewModel ViewModel = new MainPageViewModel();
```
to
```cs
public MainPageViewModel ViewModel { get; } = new MainPageViewModel();
```

This is not needed for x:Bind
```cs
this.DataContext = ViewModel;
```

**Reply by yoshiask** [reply]
The `_selectedItem` to `selectedItem` change shouldn't be necessary, the docs explicitly state support for leading underscores. Still worth a shot, of course.

**Comment by mcNets** [comment]
Take a look at [skia .csproj](https://github.com/mcNets/test_community/blob/master/test_community/test_community.Skia.Gtk/test_community.Skia.Gtk.csproj) and at [windows .csproj](https://github.com/mcNets/test_community/blob/master/test_community/test_community.Windows/test_community.Windows.csproj) both include `<PackageReference Include="CommunityToolkit.Mvvm" Version="8.0.0" />`

I will let the [project](https://github.com/mcNets/test_community) there for a while in case you want to give it a try.`

I'll try creating a new project using the dotnet command line and see what happens.

**Reply by HavenDV** [reply]
The problem here is that Uno and MVVM Toolkit both use Source Generators and there is currently no way to assign an order to call them. Therefore, the Source Generator for Uno does not see the code generated by the MVVM Toolkit. It is unlikely that there will be an opportunity for this in the near future, it is limited to dotnet itself.
You need to use a separate netstandard2.0/net6.0 library for the ViewModel, then everything will be fine.

**Reply by mcNets** [reply]
Ok, I will add a new project and give it a try.
Thanks
